### PR TITLE
Automagic dashboards improvements

### PR DESCRIPTION
**Pull request by sbelak**
# Blocking issues:
- [x] N+1 (or worse) database access. `04-05 17:28:33 DEBUG metabase.middleware :: GET /api/automagic-dashboards/database/4/candidates 200 (5 s) (10481 DB calls). Jetty threads: null/null (null busy, null idle, null queued)`
- [x] `/ready` doesn't take schema into account. We should either pick a schema at random or show shcema or otherwise do something reasonable
- [x] Server or frontend errors display a 404 page. We should convey the error to the user

## Additional TODOs
- [x] better titles
- [x] related with different depths
- [ ] support cards with aggregation
- [x] Wider selection of filters where appropriate
- [x] Make order_by work for transients (port to old format)
- [x] Add more field heuristics
- [x] limit number of cards + see more
- [x] limit number of filters
- [x] be smarter about what to use as color key
- [x] smart date resolution
- [x] If I mark a field as do not include in the data model, don't create a card
- [x] overloadable field and metric analysis
- [x] Nicer axis labels

## V2
- [ ] truncated-row fake chart type
- [ ] safe yaml
- [ ] collect dimensions with uniform distribution (count = 1 everywhere) together
- [ ] reference preexisting metrics in heuristics
- [ ] switch over to dashboard parameters

### CODE REVIEW

**Comment by sbelak** [comment]
cell dashboard
```/question/:id/cell/[base64-encoded-filter-clause]```
where base64-encoded-filter-clause is a MBQL filter expression (presumably) uniquely identifying given cell. eg:
```
[:= [:field-id 12] 1]
```

**Comment by mazameli** [comment]
Don't know if this is supposed to be working already, but it would be cool if when we see both a latitude and a longitude field that we create a pin map of the rows. (The crime stats table in our random interesting stats db is a decent example of a case where it'd be cool if we did that.)

**Comment by salsakran** [comment]
when full of tables, the div in `ready` is cut off

screen shot 2018-04-02 at 6 13 10 pm


@kdoh @tlrobinson

**Comment by salsakran** [comment]
There's also some kind of state error, when I got to `/ready/4`, it calls ` /api/automagic-dashboards/database/3/candidates` to get the candidates. On further futzing, it seems to ignore values in the url and call the candidates for DB 3.

**Comment by salsakran** [comment]
Blocking issues:
- [ ] N+1 (or worse) database access. `04-05 17:28:33 DEBUG metabase.middleware :: GET /api/automagic-dashboards/database/4/candidates 200 (5 s) (10481 DB calls). Jetty threads: null/null (null busy, null idle, null queued)`
- [ ] `/ready` doesn't take schema into account. We should either pick 
- [ ] Metric X-Rays are broken (there should be tests to verify they work)
- [ ] Field X-Rays are broken (there should be tests to verify they work)

**Comment by sbelak** [comment]
Note: `:related` currently has `:tables` disabled waiting for integration with related-entities.

**Comment by salsakran** [comment]
Seeing this from a "more" dashboard link.



screen shot 2018-04-09 at 5 36 11 pm

It only ends up being shown to the user as a 404 page
screen shot 2018-04-09 at 5 38 13 pm

**Comment by salsakran** [comment]
Fairly minor improvement to the heuristics for metric x-rays (and others)
screen shot 2018-04-09 at 5 40 02 pm

Would be much cooler if it was a set of paired tables with top 5 and bottom 5 values by each dimension.

**Comment by mazameli** [comment]
I don't think the x-ray content area should scroll independently of the rest of the page. Maybe @kdoh has an idea of what's up with that.

scroll

**Comment by mazameli** [comment]
The `zip` field in the People table is set to the Zip Code metadata type, but it seems to be behaving oddly both in automagic dashboards and in the query builder.

screen shot 2018-04-10 at 10 58 00 am

The error has to do with the Zip field not being able to be binned apparently:

screen shot 2018-04-10 at 10 59 18 am

Removing the binning results in the zip codes being distributed weirdly with skinny bars:

screen shot 2018-04-10 at 10 57 28 am

**Comment by mazameli** [comment]
Something wacky going on with the x-axis label for quarter-of-year:

screen shot 2018-04-10 at 11 14 07 am

**Comment by mazameli** [comment]
This is a fake dataset of employees, which have a Hire Date and a Birthday date field, but it's hard to tell with this scalars which date field it's using, and it's also incorrect to represent employee birthdays as "new employees." (I mean, unless they're "future employees" who've been recently born, which is more than a little dystopian.)

screen shot 2018-04-10 at 1 24 59 pm

**Comment by sbelak** [comment]
I've managed to reduce the number of DB calls to ~8/table (both for /candidates and the actual dashboard call).

File: frontend/src/metabase/components/Archived.jsx [hunk 1]
```diff
@@ -2,6 +2,9 @@ import React from "react";
 import EmptyState from "metabase/components/EmptyState";
 import Link from "metabase/components/Link";
 import { t } from "c-3po";
+
+// TODO: port to ErrorMessage for more consistent style
```
Comment (senior) [inline frontend/src/metabase/components/Archived.jsx]: Do these need to get fixed up before merge?

File: src/metabase/types.clj [hunk 2]
```diff
@@ -89,6 +90,7 @@
 
 (derive :type/CreationTimestamp :type/DateTime)
 (derive :type/JoinTimestamp :type/DateTime)
+(derive :type/Birthday :type/DateTime)
```
Comment (senior) [inline src/metabase/types.clj]: Would it be better to derive birthday from `:type/Date` since we won't have a time (I assume). Also :+1: birthday->birthdate to fit the date rhyme scheme

File: src/metabase/automagic_dashboards/rules.clj [hunk 3]
```diff
@@ -61,13 +62,13 @@
 (def ^:private FieldType (s/either (s/constrained s/Str ga-dimension?)
                                    (s/constrained s/Keyword field-type?)))
 
-(def ^:private FieldSpec (s/either [FieldType]
+(def ^:private AppliesTo (s/either [FieldType]
```
Comment (senior) [inline src/metabase/automagic_dashboards/rules.clj]: Not needed before merge, but we should switch away from `s/either` as it's deprecated.

File: src/metabase/automagic_dashboards/comparison.clj [hunk 4]
```diff
@@ -2,111 +2,119 @@
   (:require [clojure.string :as str]
```
Comment (senior) [inline src/metabase/automagic_dashboards/comparison.clj]: Is this file still needed? It looks like everything is now commented out

File: src/metabase/api/automagic_dashboards.clj [hunk 5]
```diff
@@ -35,68 +39,84 @@
 
 (api/defendpoint GET "/table/:id"
   "Return an automagic dashboard for table with id `ìd`."
-  [id]
-  (-> id Table api/check-404 magic/automagic-dashboard))
+  [id show]
```
Comment (senior) [inline src/metabase/api/automagic_dashboards.clj]: These endpoints need schemas

File: src/metabase/api/automagic_dashboards.clj [hunk 6]
```diff
@@ -35,68 +39,84 @@
...
   (-> query
       decode-base64-json
       query/adhoc-query
-      magic/automagic-analysis))
+      (magic/automagic-analysis {:show (keyword show)})))
+
+(api/defendpoint GET "/adhoc/:query/cell/:cell-query"
+  "Return an automagic dashboard analyzing ad hoc query."
+  [query cell-query show]
+  (let [query      (decode-base64-json query)
```
Comment (senior) [inline src/metabase/api/automagic_dashboards.clj]: What's going on here?

File: src/metabase/automagic_dashboards/core.clj [hunk 7]
```diff
@@ -262,27 +237,36 @@
...
-         [:dimension identifier]
-         [:aggregate-field (u/index-of #{identifier} metrics)])])))
+      [(if (dimensions identifier)
+          [:dimension identifier]
+          [:aggregation (u/index-of #{identifier} metrics)])
+        (if (= ordering "ascending")
+          :ascending
+          :descending)])))
+
+(defn merge-filters
```
Comment (senior) [inline src/metabase/automagic_dashboards/core.clj]: Would be good to reuse this instead https://github.com/metabase/metabase/blob/master/src/metabase/query_processor/middleware/expand_macros.clj#L115

File: src/metabase/automagic_dashboards/populate.clj [hunk 8]
```diff
@@ -25,28 +29,92 @@
...
+  (or (db/select-one 'Collection
+        :name     [:like "Automatically Generated Questions%"]
+        :archived false)
+      (let [c (db/count 'Collection
+                :name [:like "Automatically Generated Questions%"])]
+        (create-collection!
+         (if (pos? c)
+           (format "Automatically Generated Questions %s" (inc c))
+           "Automatically Generated Questions")
+         "#509EE3"
```
Comment (senior) [inline src/metabase/automagic_dashboards/populate.clj]: I'm not sure what this is. Probably would be easiest to def it as a constant and put a comment on it.

Comment (sbelak) [reply src/metabase/automagic_dashboards/populate.clj]: It's done like this so that the user can delete the autogenerated collection and it pops back up next time they save an xray.

File: src/metabase/automagic_dashboards/core.clj [hunk 9]
```diff
@@ -438,149 +450,220 @@
...
+                  ;; `first` realises one element at a time (no chunking).
+                  first))]
+    (do
+      (log/info (format "Applying heuristic %s to %s."
+                        (:rule rule)
+                        (:full-name root)))
+      (log/info (format "Dimensions bindings:\n%s"
+                        (->> dashboard
+                             :context
+                             :dimensions
```
Comment (senior) [inline src/metabase/automagic_dashboards/core.clj]: I'm a little confused about the usage of dimensions throughout this code. What kind of structure is it? Here it looks like it's a map, in other areas it looks like it's a list of 1 keypair maps.

File: src/metabase/automagic_dashboards/comparison.clj (hunk 4 above)
Comment (sbelak) [reply src/metabase/automagic_dashboards/comparison.clj]: It's temporary, because I kept breaking it. But maybe it would be better to pull it out until we get to it.

File: src/metabase/api/automagic_dashboards.clj (hunk 6 above)
Comment (sbelak) [reply src/metabase/api/automagic_dashboards.clj]: It gets passed in two query dicts encoded with base64. `query` is the `:dataset_query` for the card, `cell-query` is a refinement `:filter` clause that defines a cell (eg. Country = Finland)

File: src/metabase/automagic_dashboards/core.clj (hunk 9 above)
Comment (sbelak) [reply src/metabase/automagic_dashboards/core.clj]: In the heuristics it's a list of 1 keypair maps (to make the YAML at least somewhat bearable) with potential overloading. The overloading then gets resolved and dimensions get packed into a normal map. 
I'll think about better naming. Ideas welcome.

**Comment by sbelak** [comment]
Revamped naming:
* All heuristics now take an optional `transient_title` which will be used as title for transient dashboards. The existing `title` will be used when saving. If `transient_title` is omitted, `title` is used for both.
* `related` and `/candidates` now return entity name with type postfix (eg.: "Orders table", or "Created At field") as `title` and rule `:description` as `:description.
* returned dashboards now contain a new field `:transient_filters` which lists implicitly applied filters. Format: 
```
transient_filters: [
  { type: "date", field: ["Created At"], value: "March" , op: "=", :field_id 15},
  { type: "category", field: ["Product", "Category"], value: "Widget" , op: "=", field_id: 8},
]
```
When saving filters used is appended to dashboard's `:description`
* All non-table dashboards have the title in the form of: "Here's a closer look at your [[this]]"

Questions:
1) if using non-trivial filters (or, not, operators other than =), `:transient_filters` as currently shown might be confusing.
2) Since the dashboard description popup doesn't allow any formatting, filters look a bit janky when saved.
3) Do we actually need to have both title and transient_title, or can we use some consistent naming form for the saved version?

File: src/metabase/automagic_dashboards/rules.clj (hunk 3 above)
Comment (sbelak) [reply src/metabase/automagic_dashboards/rules.clj]: I know, and I hate them for it. For a large class of specs it means: go write a predicate function for the entire thing.
